lib/rust/ensogl/core/src/display/scene/layer.rs

```diff
@@ -1085,14 +1085,20 @@ impl {
     /// Decrement internal register of used [`Symbol`] instances previously instantiated with the
     /// [`instantiate`] method. In case the counter drops to 0, the caller of this function should
     /// perform necessary cleanup.
-    pub(crate) fn drop_instance<S>(&mut self, flavor: ShapeSystemFlavor) -> (usize,ShapeSystemId,PhantomData<S>)
+    pub(crate) fn drop_instance<S>(&mut self, flavor: ShapeSystemFlavor) -> (bool, ShapeSystemId, PhantomData<S>)
     where S : Shape {
-        let system_id      = ShapeSystem::<S>::id();
-        let instance_count = if let Some(entry) = self.get_mut::<S>(flavor) {
-            *entry.instance_count -= 1;
+        let system_id = ShapeSystem::<S>::id();
+        let flavor_instance_count = if let Some(entry) = self.get_mut::<S>(flavor) {
+            *entry.instance_count = entry.instance_count.saturating_sub(1);
             *entry.instance_count
         } else { 0 };
-        (instance_count,system_id,PhantomData)
+
+        let last_instance = match flavor_instance_count {
+            0 => self.total_system_instances(*system_id) == 0,
+            _ => false,
+        };
```

Comment:
I don't understand this logic. Could you please provide docs or even better, refactoring to well named variables that explain why it works this way?

Reply:
Refactored it to be hopefully a bit more clear, also added a comment. It's tricky to refactor to separate variables, since the goal is to perform the `total_system_instances` computation lazily only when necessary.